ocaml/forkexecd/lib/forkhelpers.ml

```diff
@@ -60,14 +60,32 @@ let waitpid (sock, pid) =
   | _ ->
       failwith ("This should never happen: Forkhelpers.waitpid " ^ __LOC__)
 
-let waitpid_nohang ((sock, _) as x) =
+(* [waitpid_nohang] reports the status of a socket to a process. The
+   intention is to make this non-blocking. If the process is finished,
+   the socket is closed by default and not otherwise. The [close] option
+   can be used to never close the socket such that this is left to the
+   caller, which is a cleaner protocol. *)
+let waitpid_nohang ?(close = true) (sock, pid) =
+  let fail fmt = Printf.kprintf failwith fmt in
   Unix.set_nonblock sock ;
-  let r =
-    try waitpid x
-    with Unix.(Unix_error ((EAGAIN | EWOULDBLOCK), _, _)) ->
-      (0, Unix.WEXITED 0)
-  in
-  Unix.clear_nonblock sock ; r
+  match Fecomms.read_raw_rpc sock with
+  | Fe.Finished (Fe.WEXITED n) ->
+      if close then Unix.close sock ;
+      (pid, Unix.WEXITED n)
+  | Fe.Finished (Fe.WSIGNALED n) ->
+      if close then Unix.close sock ;
+      (pid, Unix.WSIGNALED n)
+  | Fe.Finished (Fe.WSTOPPED n) ->
+      if close then Unix.close sock ;
+      (pid, Unix.WSTOPPED n)
```

Comment:
These 3 match cases could be collapsed into one:
```
| Fe.Finished ((Fe.WEXITED _ | Fe.WSIGNALED _ | Fe.WSTOPPED  _) as status) ->
   if close then Unix.close sock;
   (pid, status)
```

Reply:
Unfortunately this doesn't work because the types of Fe.WEXITED and Unix.WEXITED are not the same

Reply:
Indeed. I tried it and came back to what I had.